feat: Token search and discovery with swaps

ziad-saab opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR adds token search and discovery features to MetaMask:
- browser URL bar now allows searching for tokens by name (prefix/autocomplete) and token address
- users can swap directly from the search results (will switch or add the network if they don't have it)
- users can see asset details for searched tokens, complete with price and market data
- users can swap the found asset from the asset details page's swap button (will switch or add the network if they don't have it)

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Screenshots/Recordings**

### **After**

https://github.com/user-attachments/assets/449ff284-e127-4829-a65f-19fc9cf39a73



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Bigshmow committed:
feat: add-token-search-discovery-controller

Bigshmow committed:
chore: update package and yarn lock

Bigshmow committed:
feat: introduce to engine context

Bigshmow committed:
feat: hooks and selectors for FE

Bigshmow committed:
chore: remove empty tests

Bigshmow committed:
chore: update test json

Bigshmow committed:
chore: update failing snapshot

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
chore: add controller factory tests

Bigshmow committed:
chore: linter fix for obj property shorthand

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
chore: code coverage and resilient selector

Bigshmow committed:
chore: use requireActual method per convention and linter

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
feat: introduce generic debounce hook to searchTokens

Bigshmow committed:
Merge branch 'add-tokens-search-discovery-controller' of github.com:MetaMask/metamask-mobile into add-tokens-search-discovery-controller

Bigshmow committed:
chore: use lodash debounce directly, update test

Bigshmow committed:
feat: manage result state internally

Bigshmow committed:
chore: fix shadow

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
chore: add portfolio team to codeowners

Bigshmow committed:
chore: fix unintended notification team codeowner update

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
add tokensearchdiscoverycontroller to bg state for state change subscription

Bigshmow committed:
add controller to the get state()

Bigshmow committed:
add missing types for actions and events

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

ziad-saab committed:
feat: Token search

ziad-saab committed:
replace navigation

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

gambinish commented on app/components/Views/AssetLoader/index.tsx:
```diff
@@ -0,0 +1,55 @@
+
+    useEffect(() => {
+        if (!tokenResult) {
+            Engine.context.TokensController.addSearchedToken(address, chainId);
```
If a user searches for a token on a network not yet added to their account, we won't have some data needed to handle certain fiat conversion throughout the app (for instance, on `AsssetOverview`)

From a technical perspective, this may require us to add the required chain to the `NetworkController` state whenever you TokensController.addSearchedToken(), so that the `networkConfigurations.[chainId].nativeCurrency` for the fiat conversion is also available for the UI to consume.

We will likely also want to handle the network switch on behalf of the user, so that the correct network appears below the ticker on the TabView:

<img width="198" alt="Screenshot 2025-02-05 at 1 41 20 PM" src="https://github.com/user-attachments/assets/6ec403d5-cfca-4fae-9a63-07326eafba28" />

[Slack conversation for reference](https://consensys.slack.com/archives/C08AN1X0CSX/p1738688806325859)

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
chore: bump token controller version

Bigshmow committed:
chore: fix unnecessary format

Bigshmow committed:
chore: import from index where we can

Bigshmow committed:
chore: use testing best practices as a guideline

Bigshmow committed:
add missing discovery service

Bigshmow committed:
prefer imports from index

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
update yarn lock to use yarnpkg > jfrog

Bigshmow committed:
fix: prevent race conditions by tracking request IDs and ensuring only latest search results update state

Bigshmow committed:
add missing deps

Bigshmow committed:
fix: add mock of discovery service for tests

Bigshmow committed:
linter prefers arrow functions

Bigshmow committed:
use memo for better deps tracking

Bigshmow committed:
remove wrapper and isolate within single scope where we can

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

Bigshmow committed:
Merge branch 'main' into add-tokens-search-discovery-controller

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
add local token search controller

ziad-saab committed:
undo changes to tokens controller selectors (will be moved to discovery controller)

ziad-saab committed:
Merge remote-tracking branch 'origin/add-tokens-search-discovery-controller' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
update

ziad-saab committed:
wire up token search data to views

- fetch price from token search data controller (tsdc) when user comes from a search
- fetch market data from tsdc
- fetch token details from tsdc
- fetch swaps allowed tokens from tsdc

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
update assets-controllers local version

ziad-saab committed:
update yarn.lock

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Override Swap functionality for search & discovery assets

ziad-saab committed:
Allow user to swap assets for tokens they don't own on networks they don't have added.

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Add token search to UrlAutocomplete

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
add swap button on token search results

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
add token price and change % in search results

ziad-saab committed:
adjust assetloader styles

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Update assets-controllers

To support the search & discovery feature, updating assets-controllers to 53.1.0,
which includes `TokenSearchDiscoveryController`. This new version also has
updated peer dependencies, so we update their versions here as well.

ziad-saab committed:
remove jfrog from yarn lock

ziad-saab committed:
merge controller-utils deps in yarn lock

ziad-saab committed:
update to bogus version number

metamaskbot committed:
Bump version number to 1596

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
fix UrlAutocomplete tests

mikesposito committed:
bump controllers

mikesposito committed:
update `withKeyring` calls

mikesposito committed:
update `addNewKeyring` calls

mikesposito committed:
bump assets-controllers

mikesposito committed:
fix lint:tsc

mikesposito committed:
update lock

mikesposito committed:
remove unused ts directive

mikesposito committed:
fix unit import additional account test

ziad-saab committed:
Merge branch 'main' into mikesposito/bump-controllers

ziad-saab committed:
Merge branch 'main' into mikesposito/bump-controllers

ziad-saab committed:
Merge remote-tracking branch 'origin/mikesposito/bump-controllers' into zs/token-search-display

ziad-saab committed:
upgrade assets-controllers

ziad-saab committed:
update yarn lock

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
controllers

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
add data controller events and actions

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
remove unused import

ziad-saab committed:
remove unused import

ziad-saab committed:
Add data controller to initial background state

ziad-saab committed:
fix UrlAutocomplete tests

ziad-saab committed:
fix useTokenSearchDiscovery test

update to logic means no search if less than 2 characters typed

ziad-saab committed:
update logs test snapshot

ziad-saab committed:
update NetworkModal snapshot

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Only show network modal if we're on a search & discovery asset

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
add useAddNetwork test

ziad-saab committed:
increase coverage of useTokenSearchDiscovery tests

ziad-saab committed:
add AssetOverview tests for searched assets

ziad-saab committed:
add tsdd controller initial state

ziad-saab committed:
add AssetLoader tests

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
add token display data

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Add UrlAutocomplete tests

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
address code quality issues

sahar-fehri committed:
chore: bump assets-controllers to v53

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Only return search token results that are swappable

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
fix token search discovery tests

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
revert changes due to package updates

ziad-saab committed:
Merge remote-tracking branch 'origin/chore/bump-assets-controllers-to-v53.0.0' into zs/token-search-display

ziad-saab committed:
remove unused ts expect error

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
update yarn.lock

gambinish commented on app/components/UI/AssetOverview/TokenDetails/TokenDetails.tsx:
```diff
@@ -72,17 +73,27 @@ const TokenDetails: React.FC<TokenDetailsProps> = ({ asset }) => {
+  let conversionRate;
+  if (isAssetFromSearch(asset)) {
+    conversionRate = 1;
+  } else if (isPortfolioViewEnabled()) {
+    conversionRate = conversionRateBySymbol;
```
We should treat the `isPortfolioViewEnabled` flag as deprecated. No need to remove it here, but just a heads up that we should probably backlog a ticket for this removal across platforms.

ziad-saab replied:
@gambinish I had a feeling that this would be the case. Should I remove it and keep only the condition where it's true?

gambinish replied:
Yeah, let's treat `isPortfolioViewEnabled` as true. Condition shouldn't be necessary and we will be removing this early next week.

ziad-saab replied:
removed from TokenDetails.

gambinish commented on app/selectors/tokenSearchDiscoveryDataController.ts:
```diff
@@ -0,0 +1,28 @@
+
+export const isAssetFromSearch = (asset: unknown) => typeof asset === 'object' && asset !== null && 'isFromSearch' in asset && asset.isFromSearch === true;
+
+const selectTokenSearchDiscoveryDataControllerState = (state: RootState) =>
+  state.engine.backgroundState.TokenSearchDiscoveryDataController;
+
+export const selectTokenDisplayData = createSelector(
+  selectTokenSearchDiscoveryDataControllerState,
+  selectCurrentCurrency,
+  (_state: RootState, chainId: Hex) => chainId,
+  (_state: RootState, _chainId: Hex, address: string) => address,
+  (state, currentCurrency, chainId, address) => state?.tokenDisplayData.find(d => d.chainId === chainId && d.address === address && d.currency === currentCurrency)
+);
+
+export const selectSupportedSwapTokenAddressesByChainId = createSelector(
+  selectTokenSearchDiscoveryDataControllerState,
+  (state) => state?.swapsTokenAddressesByChainId,
+);
+
+export const selectSupportedSwapTokenAddressesForChainId = createSelector(
+  selectTokenSearchDiscoveryDataControllerState,
+  (_state: RootState, chainId: Hex) => chainId,
+  (state, chainId) => state?.swapsTokenAddressesByChainId[chainId]?.addresses,
+);
```
Can we try to convert these into `createDeepEqualSelector` so that they get memoized as much as possible?

gambinish commented on app/selectors/tokenSearchDiscoveryDataController.ts:
```diff
@@ -0,0 +1,28 @@
+import { Hex } from '@metamask/utils';
+import { selectCurrentCurrency } from './currencyRateController';
+
+export const isAssetFromSearch = (asset: unknown) => typeof asset === 'object' && asset !== null && 'isFromSearch' in asset && asset.isFromSearch === true;
```
This condition is hard for me to wrap my head around, but it seems very important to your implementation. Is there any way that we can simplify this and/or make this more readable?

At the very least I think that adding some comments here to explain what these conditions realy on would be good.

gambinish commented on yarn.lock:
```diff
@@ -4425,12 +4425,12 @@
-  version "51.0.2"
-  resolved "https://registry.yarnpkg.com/@metamask/assets-controllers/-/assets-controllers-51.0.2.tgz#fc8ed8981fce62e4c7327393faffc2017874644d"
-  integrity sha512-dEJJw6qc5HMEiMJUTOrKhTYTMS8tuf3+gofZ9PHGSvPibUx/DivC1L3bouiMNgi2xzzH9ktfmkFyCN1lzFk22w==
+"@metamask/assets-controllers@^53.1.1":
+  version "53.1.1"
+  resolved "https://registry.yarnpkg.com/@metamask/assets-controllers/-/assets-controllers-53.1.1.tgz#87e00c362247e0c0a4f3df8bd124d6e5bb0bb16c"
+  integrity sha512-AmB5/MoqCr+JJXZm0M1lmO1AUrkmSRWaVqGauBszQeXcd3aVmycZHMWEIFKNgo0bf5kX5xIkFg2CVLaMlsv+Xg==
```
We have an [open PR](https://github.com/MetaMask/metamask-mobile/pull/14201) to upgrade AssetsController to 55.0.1.

Is this patch still necessary? Or can we use the bumped version?

ziad-saab replied:
This is not necessary. I brought in the changes from the open PR so that more tests would pass. Once the open PR to upgrade packages is merged, I expect all these changes to not be in my PR anymore.

This is important: I don't want this PR to contain any updates to packages. I'd rather the package updates be done in a separate PR, so that in case this current PR needs to be reverted, it won't affect the upgrade of packages.

ziad-saab replied:
patch removed due to assets-controllers update to v55 from another PR

gambinish commented on app/components/hooks/useAddNetwork.tsx:
```diff
@@ -0,0 +1,45 @@
+        setRejectAddNetwork(noop);
+    };
+
+    const networkModal = !popularNetwork ? null : (
+         <NetworkModals
+            isVisible
+            navigation={navigation}
+            onClose={onCloseModal}
+            networkConfiguration={popularNetwork}
+            showPopularNetworkModal
+            shouldNetworkSwitchPopToWallet={false}
+            autoSwitchNetwork
+            onNetworkSwitch={noop}
+            onAccept={resolveAddNetwork}
+            onReject={rejectAddNetwork}
+         />
+    );
```
Do we need to return the network modal jsx in this hook? I've run into difficult to troubleshoot rendering bugs with this pattern in the past.

My preference would be to return a boolean value, such as `shouldShowNetworkModal`, and read the hook from a react component, and then conditionally render the JSX in the component's body accordingly.

ziad-saab replied:
@gambinish I'd love to hear more about the rendering bugs you experienced in the past. In this case here, note the reliance on `onAccept` and `onReject` which are defined in the hook. If we were to just use a flag, we'd have to expose `resolveAddNetwork` and `rejectAddnetwork`.

It's not impossible nor is it even a problem, but I'd like to find out more about the bugs you experienced in the past to be able to make the best decision about this. I agree that it's not a common pattern. I picked it because it solved the problem with as much encapsulation as it could.

gambinish replied:
While this technically works, the main anti-pattern here is that hooks are meant to run inside of React component, not behave like them. This can sometimes lead to unpredictable behavior, especially when other JSX elements are calling hooks of their own. Hooks were meant to separate logic from UI, rather than co-mingling them.

What would happen if we eventually needed to add some custom hook inside of `NetworkModals`? Then any component consuming `useAddNetwork` would not only call your hook, but may result in multiple render trees that are hard to track or debug.

There are also some testing implications here since you'd now need to test rendering inside of a hook.

ziad-saab replied:
Hey @gambinish thanks for trying to explain the issue! I'm still having trouble grasping it. I'd really love to take advantage of this issue to upgrade my React skills and understanding.

> What would happen if we eventually needed to add some custom hook inside of NetworkModals? Then any component consuming useAddNetwork would not only call your hook, but may result in multiple render trees that are hard to track or debug.

My understanding is that `useAddNetwork` is simply a function returning a piece of JSX to be rendered inside another component. Only when that piece of JSX gets rendered do its hooks take effect. I'd like to understand the difference between returning this piece of JSX (basically a call to `React.createElement` the way I understand it) from `useAddNetwork`, and creating it inside another component.

> There are also some testing implications here since you'd now need to test rendering inside of a hook.

I'd love to understand the implications. This would not only improve my React-fu, but it would help me create a better solution than the one I have here.

gambinish replied:
If values like `resolveAddNetwork` change between renders, the JSX inside the hook can end up referencing stale versions of those values. Since that JSX is being re-created on every render, it can also trigger unnecessary re-renders of the modal, especially if multiple components are using the hook and each one is rendering its own instance. It just adds up fast and makes things harder to manage.

As far as testing goes, when the hook just returns data and functions (instead of JSX), you can test it in isolation without having to render anything. No need for snapshot tests or mounting modals — you just assert that the hook gives you the right state and behaviors.

I’d recommend returning just the state and control functions from the hook, and rendering `NetworkModals` in the parent component instead:

```
const { addPopularNetwork, showNetworkModal, onCloseModal, ... } = useAddNetwork();

return (
  <>
    {showNetworkModal && (
      <NetworkModals
        isVisible
        navigation={navigation}
        onClose={onCloseModal}
        networkConfiguration={popularNetwork}
        showPopularNetworkModal
        shouldNetworkSwitchPopToWallet={false}
        autoSwitchNetwork
        onNetworkSwitch={noop}
        onAccept={resolveAddNetwork}
        onReject={rejectAddNetwork}
      />
    )}
  </>
);
```

Added benefit is that this keeps your logic reusable. The hook can be used anywhere, but the UI doesn’t matter. So, if you want to use the hook, but render a different modal, or even the same modal with different props, you can do so without having to untangle the UI from the hook.

gambinish replied:
I also don't want to be a blocker. If you feel like this change is not reasonable, let me know

ziad-saab replied:
@gambinish thanks a lot for these insights! How about, instead of a hook, if this was a component with a `useImperativeHandle` and a ref, then we can call `await addNetworkRef.current?.addNetwork(network)`?

ziad-saab replied:
Following our conversation on Slack, we'll keep it as a hook. It'll be converted to a component in the next PR for token discovery.

gambinish commented on app/components/Views/BrowserTab/BrowserTab.tsx:
```diff
@@ -1175,10 +1175,18 @@ export const BrowserTab: React.FC<BrowserTabProps> = ({
    */
-  const onSelect = (url: string) => {
+  const onSelect = (item: AutocompleteSearchResult) => {
     // Unfocus the url bar and hide the autocomplete results
     urlBarRef.current?.hide();
-    onSubmitEditing(url);
+
+    if (item.type === 'tokens') {
+      navigation.navigate(Routes.BROWSER.ASSET_LOADER, {
+        chainId: item.chainId,
+        address: item.address,
+      });
+    } else {
+      onSubmitEditing(item.url);
+    }
```
Let's convert this into a `useCallback` hook to prevent unnecessary re-renders.

gambinish commented on app/components/Views/AssetLoader/index.tsx:
```diff
@@ -0,0 +1,58 @@
+    const { styles } = useStyles(styleSheet, {});
+
+    useEffect(() => {
+        Engine.context.TokenSearchDiscoveryDataController.fetchTokenDisplayData(chainId, address);
```
```suggestion
const fetchTokenDisplayData = Engine.context.TokenSearchDiscoveryDataController.fetchTokenDisplayData
fetchTokenDisplayData(chainId, address);
```
```suggestion
        Engine.context.TokenSearchDiscoveryDataController.fetchTokenDisplayData(chainId, address);
```

ziad-saab replied:
@gambinish thanks for this suggestion! I'd like to hear more about what prompted it, so that it may inform further coding decisions.

gambinish replied:
This is just a readability nit, doesn't need to block

gambinish reviewed (changes requested):
Nice job! I left a few comments on things that caught my eye. Feel free to push back where you feel I may have misunderstood something.

Additionally, it looks like we need to `lint:tsc` and get Bitrise to pass. Could you update this branch with `main` and then re-run Bitrise? Once it's fully green I'm happy to pull this and run it locally.

Feel free to ping me directly on Slack as well.

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
convert token search selectors to createDeepEqual

ziad-saab committed:
add explanation of isAssetFromSearch

ziad-saab commented:
@gambinish thanks for the review! The `lint:tsc` errors are due to some package updates that should not even be part of this PR. I'm waiting for the package update PRs to be merged first, then I'll merge `main` into this PR. At that point there should be less files changed, more code coverage, and no more errors.

ziad-saab committed:
convert some functions to useCallback in BrowserTab

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
bogus version number

metamaskbot committed:
Bump version number to 1664

metamaskbot committed:
Bump version number to 1665

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Remove isPortfolioView from token details

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
restore ts-expect-error comments

metamaskbot committed:
Bump version number to 1667

ziad-saab committed:
fix lint errors

ziad-saab committed:
update build version

metamaskbot committed:
Bump version number to 1668

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
update browser bar placehholder text

ziad-saab committed:
update snapshots

ziad-saab committed:
revert version number changes

ziad-saab committed:
Merge branch 'main' into zs/token-search-display

darkwing committed:
fix: Add feature flag for token search (#14599)

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
fix typescript errors

ziad-saab committed:
add missing dep to useMemo

ziad-saab committed:
update changelog

ziad-saab committed:
add feature flag to use token search hook

ziad-saab committed:
fix mock state for use token search hook test

smilingkylan reviewed (commented):
FYI I'm not seeing anything my browser search results when I build your branch. Do I need to enable / disable a feature flag or anything like that?
![image](https://github.com/user-attachments/assets/1abdd7fe-cf18-4934-ae1e-bbb0bc7d5654)

smilingkylan commented on app/components/UI/UrlAutocomplete/Result.tsx:
```diff
@@ -6,61 +6,116 @@ import WebsiteIcon from '../WebsiteIcon';
     const styles = stylesheet({theme});
 
-    const name = typeof result.name === 'string' ? result.name : getHost(result.url);
+    const name = typeof result.name === 'string' || result.type === 'tokens' ? result.name : getHost(result.url);
```
Could you please change the property name from `type` to something that is more easily searchable? My concern here is that if someone tries to search for this property they will end up with thousands of matches in this repo

smilingkylan commented on app/components/UI/AssetOverview/AssetOverview.tsx:
```diff
@@ -333,15 +375,21 @@ const AssetOverview: React.FC<AssetOverviewProps> = ({
-      ? exchangeRate * tickerConversionRate
-      : 0;
+  if (isAssetFromSearch(asset) && tokenResult?.found) {
+    currentPrice = tokenResult.price?.price || 0;
```
IMO having a property like `price` that is nested inside of another property named `price` will cause confusion. If you have the ability to adjust the parent or child property name then I would like it changed

ziad-saab replied:
These results are coming from an API and through the `assets-controller`. Changing this would entail a bunch of delays to change the `assets-controller` package, get it released, and then upgraded in mobile. Are we able to live with this?

smilingkylan replied:
@ziad-saab Re-reviewing right now

No need to delay. I will convey this concern to the assets team(?) so they can keep it in mind going forward.

smilingkylan commented on app/components/UI/UrlAutocomplete/index.tsx:
```diff
@@ -40,25 +55,48 @@ const UrlAutocomplete = forwardRef<
+  }, [currentCurrency]);
+
+  const resultsByCategory: {category: string, data: AutocompleteSearchResult[]}[] = useMemo(() => (
+    ORDERED_CATEGORIES.flatMap((category) => {
```
`category` is a better keyname for the types of results, IMO

smilingkylan commented on app/components/UI/UrlAutocomplete/index.tsx:
```diff
@@ -40,25 +55,48 @@ const UrlAutocomplete = forwardRef<
+
+  const resultsByCategory: {category: string, data: AutocompleteSearchResult[]}[] = useMemo(() => (
+    ORDERED_CATEGORIES.flatMap((category) => {
+      if (category === 'tokens') {
```
We may also want to start using a constant / enum for the types rather than a hard-coded string

smilingkylan commented on app/components/UI/UrlAutocomplete/index.tsx:
```diff
@@ -148,21 +204,90 @@ const UrlAutocomplete = forwardRef<
+    hide();
+    onDismiss();
+    navigation.navigate('Swaps', {
+      screen: 'SwapsAmountView',
```
Consider using constants here for `'Swaps` and `SwapsAmountView`

smilingkylan commented on app/components/hooks/useTokenSearchDiscovery/useTokenSearchDiscovery.ts:
```diff
@@ -7,8 +7,15 @@ import {
+import { tokenSearchDiscoveryEnabled } from '../../../selectors/featureFlagController/tokenSearchDiscovery';
 
-const SEARCH_DEBOUNCE_DELAY = 300;
+const SEARCH_DEBOUNCE_DELAY = 50;
```
Why have you reduced the debounce here? 0.05 seconds is very quick

ziad-saab replied:
To get a more reactive UI. @smilingkylan I wonder if there's a happy value between 50 and 300 that wouldn't feel too quick to you, while also showing results promptly...

smilingkylan replied:
@ziad-saab My concern is that 50ms is almost unnoticeable and at that point there is minimal reason for even having a debounce. I say set it to 200ms and if it feels good then go with that. Should still be relatively snappy.

ziad-saab replied:
How do you feel about 150? :-)

smilingkylan commented on app/components/Views/Asset/index.js:
```diff
@@ -521,7 +523,7 @@ class Asset extends PureComponent {
       asset.isETH ||
       asset.isNative ||
-      asset.address?.toLowerCase() in this.props.swapsTokens;
+      (isAssetFromSearch(asset) ? this.props.searchDiscoverySwapsTokens?.includes(asset.address?.toLowerCase()) : asset.address?.toLowerCase() in this.props.swapsTokens);
```
Consider adding improving the readability of this conditional

smilingkylan reviewed (dismissed):
Requested some minor changes. Otherwise PR seems easy to read, so thank you for that

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

smilingkylan commented:
> FYI I'm not seeing anything my browser search results when I build your branch. Do I need to enable / disable a feature flag or anything like that?

@ziad-saab Looks like it's most likely a result of the TokenSearchDiscoveryController not having been updated yet.

ziad-saab committed:
- Change property name `type` to `category` for search results
- Change token search endpoint to swappable token search

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Update token search controller to latest version

ziad-saab committed:
remove jfrog

ziad-saab committed:
fix urlautocomplete test

ziad-saab committed:
fix jittery animation on navigating to asset from search

ziad-saab committed:
use enum for autocomplete search result categories

ziad-saab committed:
use constants for route and screen names in urlautocomplete

ziad-saab committed:
improve readability of complex nested ternary

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
fix UrlAutocomplete tests

ziad-saab committed:
update AssetDetailsActions snapshot

ziad-saab committed:
fix WalletActions tests

gambinish commented on app/components/UI/UrlAutocomplete/index.test.tsx:
```diff
@@ -5,8 +5,105 @@ import { act, fireEvent, screen } from '@testing-library/react-native';
+  reset: () => void;
+  searchTokens: () => void;
+}) => {
+  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+  const useTSD = require('../../hooks/useTokenSearchDiscovery/useTokenSearchDiscovery');
```
Can we convert this to use an ESM import instead?

gambinish reviewed (dismissed):
Approving with a small comment. Happy to approve this in a fast follow up PR to not block this feature.

ziad-saab committed:
increase debounce delay

smilingkylan reviewed (dismissed):
LGTM 👍 with one caveat: I need confirmation that other devs are able to get the HTTP requests for the token search working

darkwing commented:
Everything loading great for me!

<img width="430" alt="SCR-20250421-nzyz" src="https://github.com/user-attachments/assets/f4c9c316-8091-44fb-869d-2e3b5f47bd90" />

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display

ziad-saab committed:
Added GHSA-h9w6-f932-gq62 to .iyarc

According to this thread https://consensys.slack.com/archives/C010Y3K38PM/p1745313131151649
we are not affected by it

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/token-search-display
